core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1581,6 +1650,178 @@ private void validateTableModelAndCreateTypeCast(
+            }
+            path.of(configuration.getBackupRoot()).concat(configuration.getBackupTempDirName()).put(Files.SEPARATOR).$();
+            cachedTmpBackupRoot = path.toString();
+            if (!ff.exists(path)) {
+                LOG.info().$("Creating backup directory ").$(cachedTmpBackupRoot).$();
+                if (ff.mkdirs(path, configuration.getMkDirMode()) != 0) {
+                    throw CairoException.instance(ff.errno()).put("Could not create [dir=").put(path).put(']');
+                }
+            }
+        }
+
+        CairoSecurityContext securityContext = executionContext.getCairoSecurityContext();
+        try (TableReader reader = engine.getReader(securityContext, tableName)) {
+            cloneMetaData(tableName, configuration.getRoot(), cachedTmpBackupRoot, configuration.getMkDirMode(), reader);
+            try (TableWriter backupWriter = engine.getBackupWriter(securityContext, tableName, cachedTmpBackupRoot)) {
+                RecordCursor cursor = reader.getCursor();
+                copyTableData(cursor, reader.getMetadata(), backupWriter);
+                backupWriter.commit();
+            }
+        }
+
+        path.of(configuration.getBackupRoot()).concat(configuration.getBackupTempDirName()).put(Files.SEPARATOR).concat(tableName).$();
+        int renameRootLen = renamePath.length();
+        try {
+            renamePath.trimTo(renameRootLen).concat(tableName).$();
+            if (!ff.rename(path, renamePath)) {
+                throw CairoException.instance(ff.errno()).put("Could not rename [from=").put(path).put(", to=").put(renamePath).put(']');
+            }
+            LOG.info().$("Completed backup of ").$(tableName).$(" to ").$(renamePath).$();
+        } finally {
+            renamePath.trimTo(renameRootLen).$();
+        }
+    }
+
+    private void cloneMetaData(CharSequence tableName, CharSequence root, CharSequence backupRoot, int mkDirMode, TableReader reader) {
+        path.of(backupRoot).concat(tableName).put(Files.SEPARATOR).$();
+
+        if (ff.exists(path)) {
+            throw CairoException.instance(0).put("Backup dir for table \"").put(tableName).put("\" already exists [dir=").put(path).put(']');
+        }
+
+        if (ff.mkdirs(path, mkDirMode) != 0) {
+            throw CairoException.instance(ff.errno()).put("Could not create [dir=").put(path).put(']');
+        }
+
+        TableReaderMetadata sourceMetaData = (TableReaderMetadata) reader.getMetadata();
+        int rootLen = path.length();
+        try {
+            mem.of(ff, path.trimTo(rootLen).concat(TableUtils.META_FILE_NAME).$(), ff.getPageSize());
+            sourceMetaData.cloneTo(mem);
+
+            // create symbol maps
+            path.trimTo(rootLen).$();
+            int symbolMapCount = 0;
+            for (int i = 0; i < sourceMetaData.getColumnCount(); i++) {
```

Comment:
sorry to be a pain, all other "for" loops across the project at limited by constant rather than function. The rationale is to avoid calling function in a loop unnecessarily (which java will do).

This may not matter as much here, but still it goes on the assumption that `sourceMetaData.getColumnCount()` is lightweight, which it is, but only coincidentally.

More robust loop would be:

```java
for (int i = 0, n = sourceMetaData.getColumnCount(); i < n; i++) {
```